### ISSUE TITLE
Performance regression in fms-hf-tuning image

### ISSUE THREAD

**Issue by kpouget** [foundation-model-stack/fms-hf-tuning#201]
## Describe the bug

As part of our performance evaluation of the fms-hf-tuning fine-tuning stack, we observed a regression between the images we tested:

*    quay.io/modh/fms-hf-tuning:01b3824c9aba22d9d0695399681e6f0507840e7f --> fast (April 17)
```
{'train_runtime': 829.2865, 'train_samples_per_second': 62.707, 'train_steps_per_second': 3.919, 'train_tokens_per_second': 12141.512, 'train_loss': 2.287814903846154, 'epoch': 1.0}
```
*    quay.io/modh/fms-hf-tuning:bd8bf628cd739c7a201a976bc3c1096785353f1a --> slow (May 26)
```
{'train_runtime': 1391.6176, 'train_samples_per_second': 37.368, 'train_steps_per_second': 2.335, 'train_tokens_per_second': 7235.315, 'train_loss': 2.287858623798077, 'epoch': 1.0}
```
*    quay.io/modh/fms-hf-tuning:84b0337b7baee119e909d4e901b6dadfe34c1f9a --> slow (May 21, planned for delivery)
```
{'train_runtime': 1404.3311, 'train_samples_per_second': 37.03, 'train_steps_per_second': 2.314, 'train_tokens_per_second': 7169.813, 'train_loss': 2.2878201622596155, 'epoch': 1.0}
```
Additionally, we observed that with the bd8bf628cd739c7a201a976bc3c1096785353f1a image, the multi-model tuning performance (multiple independent fine-tuning jobs, running on a dedicated GPU) are degraded compared to the reference run (one job running alone in the cluster). This wasn't the case in our first tests with 84b0337b7baee119e909d4e901b6dadfe34c1f9a.

* fine tuning of gpt_bigcode-santacoder against the Alpaca dataset

## Platform

OpenShift AI

## Steps to reproduce

1. Create these 3 objects in a namespace,
    * [pytorchjob.yaml.log](https://github.com/user-attachments/files/15900894/pytorchjob.yaml.log)
    * [cm_entrypoint.yaml.log](https://github.com/user-attachments/files/15900895/cm_entrypoint.yaml.log)
    * [cm_config.yaml.log](https://github.com/user-attachments/files/15900896/cm_config.yaml.log)
2. See the Pod being executed with the `quay.io/modh/fms-hf-tuning:release` image, currently hitting the regression
3. When the Pod finishes, create an interactive Pod with root privileges (to be able to install the new package)
```
oc debug fine-tuning-master-0 --as-root
```
4. Prepare the environment, by running the script a first time
```
$ bash /mnt/entrypoint/entrypoint.sh
+ grep transformers
transformers==4.41.2
...
{'train_runtime': 146.0027, 'train_samples_per_second': 35.616, 'train_steps_per_second': 2.226, 'train_tokens_per_second': 6719.494, 'train_loss': 2.377412109375, 'epoch': 1.0}
```
5. install the 4.40.2 version of transformers and tell Python to use it
```
export HOME=/tmp
pip install transformers==4.40.2 --user
export PYTHONPATH=/tmp/.local/lib/python3.11/site-packages/:$PYTHONPATH
```
6. Run the script again, with the 4.40 version of `transformers`
```
$ bash /mnt/entrypoint/entrypoint.sh
+ grep transformers
transformers==4.40.2
...
{'train_runtime': 88.4812, 'train_samples_per_second': 58.77, 'train_steps_per_second': 3.673, 'train_tokens_per_second': 11087.825, 'train_loss': 2.377412109375, 'epoch': 1.0}
```
7. **Compare the `train_runtime` and `train_samples_per_second` etc and notice the difference.**

Note that:
* you can configure the length of the dataset with `DATASET_REPLICATION`. Set to `0.1` to get results faster.
* the version pinning has been tested without the fiddling above, only required to make a simple reproducer.

## Expected behavior

![image](https://github.com/foundation-model-stack/fms-hf-tuning/assets/7559202/675e09ee-ca20-4779-aa1a-dcba8d5708ed)

## Observed behavior

![image](https://github.com/foundation-model-stack/fms-hf-tuning/assets/7559202/57f99e86-4d97-478f-af86-6d6de9f0f188)

## See also

[RHOAIENG-8551](https://issues.redhat.com/browse/RHOAIENG-8551)

**Comment by alex-jw-brooks** [comment]
Hey @kpouget, thanks for raising this issue and for providing the kube configs, it was super helpful! I did some investigation in the `quay.io/modh/fms-hf-tuning:release` image and had some findings that were similar, and some that were a bit surprising.

The short story is that I was able to reproduce your findings for `4.41.2` being slower using the same dataset & 0.1 for dataset replication. However, it also seems like the speed of the tuning is inconsistent. For my tunings, I was generally seeing:

`4.40.2` - most trainings took 500-600 seconds, and generally sat around 550. 

`4.41.2` - some trainings took ~700+ seconds. However, some ran way faster (100-150) seconds, where all trainings here leveraged the same config, and produced the same loss. E.g.,
```
{'train_runtime': 692.3355, 'train_samples_per_second': 7.511, 'train_steps_per_second': 0.469, 'train_tokens_per_second': 1417.036, 'train_loss': 2.377412109375, 'epoch': 1.0}

{'train_runtime': 109.6998, 'train_samples_per_second': 47.402, 'train_steps_per_second': 2.963, 'train_tokens_per_second': 8943.169, 'train_loss': 2.377412109375, 'epoch': 1.0}
```

In all experiments that I ran, the tuning was running in a same pod and doing the data formatting etc wit the same mounts the pods spawned with the custom resource leverage. I'm curious if there is some caching at work here, but it is strange -  I should also mention that the time per step seems pretty consistent in the progress bar outputs once things have actually started.

I did some investigation into the concurrency issues as well, and was not able to produce any issues there with either version of transformers - I did see the same weird behavior from `4.41.2` where some runs were going faster than others, but it didn't seem to depend on jobs running in the cluster concurrently, as I saw slow/fast runs with `4.41.2` when just one job was running also.

After discussing with @anhuong / @Ssukriti, we've decided to pin to `4.40` until the reasons for these inconsistent timings has been more thoroughly investigated

**Comment by Ssukriti** [comment]
Thank you Alex! 
Michael's team will be investigating further to understand impact over larger set of tests with granite models. 

There was a request to get full set of dependencies in the latest RH image to verify other dependencies besides transformers. 

The RH image mentioned above is `quay.io/modh/fms-hf-tuning:84b0337b7baee119e909d4e901b6dadfe34c1f9a` -> @anhuong could you pull out dependency list from this image and attach it here? Alternatively you could use any of our internal images that correspond to 0.3 release and get list as well. Thank you!

**Comment by anhuong** [comment]
`quay.io/modh/fms-hf-tuning:84b0337b7baee119e909d4e901b6dadfe34c1f9a` was built a month ago, May 31st.
pip freeze deps:
```
accelerate==0.30.1
aiohttp==3.9.5
aiosignal==1.3.1
attrs==23.2.0
certifi==2024.2.2
charset-normalizer==3.3.2
datasets==2.19.1
dill==0.3.8
docstring_parser==0.16
einops==0.8.0
filelock==3.14.0
fire==0.6.0
flash-attn==2.5.9.post1
fms-hf-tuning @ file:///tmp/fms_hf_tuning-0.1.dev1%2Bg84b0337.d20240531-py3-none-any.whl#sha256=fbb236a382849683a7fd49fd9e341440652b08eaf0353edee052952de7dab696
frozenlist==1.4.1
fsspec==2024.3.1
huggingface-hub==0.23.2
idna==3.7
Jinja2==3.1.4
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
mpmath==1.3.0
multidict==6.0.5
multiprocess==0.70.16
networkx==3.3
numpy==1.26.4
nvidia-cublas-cu12==12.1.3.1
nvidia-cuda-cupti-cu12==12.1.105
nvidia-cuda-nvrtc-cu12==12.1.105
nvidia-cuda-runtime-cu12==12.1.105
nvidia-cudnn-cu12==8.9.2.26
nvidia-cufft-cu12==11.0.2.54
nvidia-curand-cu12==10.3.2.106
nvidia-cusolver-cu12==11.4.5.107
nvidia-cusparse-cu12==12.1.0.106
nvidia-nccl-cu12==2.20.5
nvidia-nvjitlink-cu12==12.5.40
nvidia-nvtx-cu12==12.1.105
packaging==24.0
pandas==2.2.2
peft==0.11.1
psutil==5.9.8
pyarrow==16.1.0
pyarrow-hotfix==0.6
Pygments==2.18.0
python-dateutil==2.9.0.post0
pytz==2024.1
PyYAML==6.0.1
regex==2024.5.15
requests==2.32.3
rich==13.7.1
safetensors==0.4.3
sentencepiece==0.2.0
shtab==1.7.1
simpleeval==0.9.13
six==1.16.0
sympy==1.12.1
termcolor==2.4.0
tokenizers==0.19.1
torch==2.3.0
tqdm==4.66.4
transformers==4.41.2
triton==2.3.0
trl==0.8.6
typing_extensions==4.12.0
tyro==0.8.4
tzdata==2024.1
urllib3==2.2.1
xxhash==3.4.1
yarl==1.9.4
```

**Comment by michael-johnston** [comment]
Reporting here results of some tests on this issue that we ran last week

## What we tested

* Model: llama3-8b
* Method: full fine-tune
* GPU Models: A100-80GB, L40S
* (fms-hf-tuning commit,transformers version): (db99b280m, 4.38.2) AND (087e0c47, 4.41.2)

We tested a range of batch sizes and tokens-per-sample using our throughput benchmark experiment

## What we found

Summary is that the v4.41.2 transformers version
*  Improves memory utilisation significantly (up to 32GB less memory consumption) when total tokens per batch 16384 or less on all points tested. This allows e.g. full fine-tune of llama3-8b to work on 2 L40S where previously could not 
* Causes an average degradation of 17% on throughput **on 2 A100-80GB** across all (batch-size, tokens-per-sample) combinations 
* For 4 A100s and 4 L40S the throughput of individual points can improve or degrade but on average there is no impact. 

There is no discernible correlation between the throughput degradation and memory utilisation improvement at individual points. That is  throughput degradation occurs at points where memory utilisation does not change, and points where memory utilisation decreases still exhibit similar throughput. However, it's certainly tempting to speculate that a change improving memory efficiency has had impacted throughput in certain situations. 

Details follow

### 2 A100s

1. Performance with transformers v4.41.2 is degraded w.r.t v4.38.2 on 2 A100 GPUs on **all** (batch-size, tokens-per-sample) combinations tested, by an average of 17% and maximum of 50% (dataset-tokens-per-second)
2. GPU memory utilisation with transformers v4.41.2  decreases by 16GB on average on 2 A100s. The maximum decrease was 32GB. The saving is when the total tokens per batch is 16384 or less.

### 4 A100s

1. Performance with transformers v4.41.2 is overall similar between 4.38.2 on 4 A100 and 4 L40S but there are changes across all points tested. The mean difference is -0.7% (new version is faster on average). The max degradation is 8%, the max improvement is 15%
2. GPU memory utilisation decreases by ~8GB on average with the new version. The maximum decrease was 18 GB. The saving is when the total tokens per batch is 16384 or less

### 2 L40S

1. Increase in memory efficiency allows full fine-tune of llama3-8b on 2xL40S where before it was not possible (when total tokens per batch is 2048 or less)

**Comment by chakrn** [comment]
@kpouget Amita said in the call with SW this morning that this was still an issue for the 4.42 transformers. Can you post the updated performance numbers you're getting with 4.42?

**Comment by kpouget** [comment]
Hello Neil and all, yes, I was finalizing the measurements this week with transformers 4.42, and how it compares with 4.40 (best so far) and 4.41 (worst so far)

Job runtime

![image](https://github.com/user-attachments/assets/a33b7295-f712-4e42-9975-28cb2053c900)
![image](https://github.com/user-attachments/assets/541fde9b-3cda-46ad-ba0c-70f96c0149dc)
![image](https://github.com/user-attachments/assets/a0179393-55e9-47d1-8798-d2c4001ddce8)

GPU total memory usage: 4.41 consumes slightly less GPU memory

![image](https://github.com/user-attachments/assets/ded5c4cd-5580-4b34-b1a6-6a33785da883)
![image](https://github.com/user-attachments/assets/83a37840-1c40-4bbc-b6ba-42e2b26fd8e8)
![image](https://github.com/user-attachments/assets/e98d02e9-8578-4360-9a0d-093c093c7a50)

Multi-model testing: compare 1 job running standalone VS 8 jobs running concurrently

![image](https://github.com/user-attachments/assets/09128606-d080-4384-bc9f-3a11ef0ae8c5)
![image](https://github.com/user-attachments/assets/8ae32603-1fe3-4d85-a2f8-30c0aa93387f)
![image](https://github.com/user-attachments/assets/d23431f6-fa47-44f7-9a50-3fe9479b092b)
![image](https://github.com/user-attachments/assets/d5176b5d-d009-44ce-a084-844313e48efe)

**Comment by praveenj83** [comment]
Thanks @kpouget for sharing these results.

@michael-johnston Is this in line with what you have seen as well? I think this is fine as it is more important for us to stay abreast with the latest and transformers 4.43 is going to have some of the substantial throughput improvements we have contributed.

**Comment by Ssukriti** [comment]
conclusion is : 
1. we have skipped any release image containing transformers 4.41 due to major degradation caught by RH . Openshift 2.11 has transformers 4.40
2. Latest release has transformers 4.42 , which will be shipped in openshift 2.12 by RH . We have found that to be acceptable performance. 
3. Next release for openshift 2.13 will have throughput improvements from transformers 4.43

Hence the issue can be closed , unless anyone has objections
